libyul/YulString.h

```diff
@@ -43,31 +43,38 @@ class YulStringRepository: boost::noncopyable
 		std::uint64_t hash;
 	};
 	YulStringRepository():
-		m_strings{std::make_shared<std::string>()},
+		m_strings{std::make_shared<std::pair<std::string, size_t>>()},
 		m_hashToID{std::make_pair(emptyHash(), 0)}
 	{}
 	static YulStringRepository& instance()
 	{
 		static YulStringRepository inst;
 		return inst;
 	}
-	Handle stringToHandle(std::string const& _string)
+	Handle stringToHandle(std::string const& _string, const size_t _suffix)
 	{
-		if (_string.empty())
+		if (_string.empty() && _suffix == 0)
 			return { 0, emptyHash() };
-		std::uint64_t h = hash(_string);
+		std::uint64_t h = hash(_string, _suffix);
 		auto range = m_hashToID.equal_range(h);
 		for (auto it = range.first; it != range.second; ++it)
-			if (*m_strings[it->second] == _string)
+			if (m_strings[it->second]->first == _string && m_strings[it->second]->second == _suffix)
 				return Handle{it->second, h};
-		m_strings.emplace_back(std::make_shared<std::string>(_string));
+		m_strings.emplace_back(std::make_shared<std::pair<std::string, size_t>>(std::make_pair(_string, _suffix)));
 		size_t id = m_strings.size() - 1;
 		m_hashToID.emplace_hint(range.second, std::make_pair(h, id));
 		return Handle{id, h};
 	}
-	std::string const& idToString(size_t _id) const	{ return *m_strings.at(_id); }
-
-	static std::uint64_t hash(std::string const& v)
+	std::string const idToString(size_t _id) const
+	{
+		return m_strings[_id]->second > 0 ? m_strings[_id]->first + "_" + std::to_string(m_strings[_id]->second)
+										  : m_strings[_id]->first;
+    }
+	std::string const idToPrefix(size_t _id) const
```

Comment:
This can and should return by reference (unless we add the suffix to the string after all as suggested in a different comment).